howso/client/schemas/reaction.py

```diff
@@ -35,313 +150,179 @@ class Reaction(abc.MutableMapping):
-        details : dict of list
-            The reaction details.
-
-        Returns
-        -------
-        List of dicts
-            One dict per case.
+        reactions : list of Reaction
+            One or more Reaction objects to accumulate to this Reaction.accumulate
         """
-        if isinstance(details, list):
-            return details
-
-        if details in [None, np.nan]:
-            return []
-
-        # Ensure only "known" keys are present in the details.
-        if extra_keys := set(details.keys()) - cls.KNOWN_KEYS - cls.SPECIAL_KEYS:
-            warnings.warn(
-                f'Unrecognized detail keys found: [{", ".join(extra_keys)}] '
-                f'and ignored.')
-        cleaned_details = {
-            k: v for k, v in details.items()
-            if k in cls.KNOWN_KEYS and v
-        }
-        # Transform dict[list] -> list[dict]
-        per_case_details = [
-            dict(zip([key for key in cleaned_details.keys()], values))
-            for values in zip(*cleaned_details.values())
-        ]
-        # Re-insert the constant, special_keys and their values.
-        for case in per_case_details:
-            case.update({key: details[key] for key in cls.SPECIAL_KEYS})
-
-        return per_case_details
-
-
-# Doesn't work if this is inside the `Reaction` class.
-@Reaction.add_reaction.register
-def _(self, reaction: "Reaction"):
-    """Add another `Reaction` to Reaction."""
-    return self.add_reaction(reaction.get("action"), reaction.get("details", {}))
+        if not isinstance(reactions, Sequence):
+            reactions = [reactions]
+        for reaction in reactions:
+            if not isinstance(reaction, Reaction):
+                raise TypeError(f"All items in `reactions` must be of type `Reaction` (found type: {type(reaction)}).")
+            if reaction["action"] is not None:
+                if self._action is not None:
+                    self._action = pd.concat([self._action, reaction["action"]])
+                else:
+                    self._action = reaction["action"]
+
+            if self._details is not None:
+                for key, detail in reaction["details"].items():
+                    if detail is None:
+                        continue
+                    elif key not in self._details:
+                        self._details[key] = detail
+                    elif hasattr(detail, "extend") and callable(detail.extend):
+                        self._details[key].extend(detail)
+                    elif isinstance(detail, pd.DataFrame):
+                        self._details[key] = pd.concat([self._details[key], detail])
+                    else:
+                        raise TypeError(
+                            f"The value under the key {key} was expected to be a list (or another "
+                            f"MutableSequence) or DataFrame but it is of type "
+                            f"{type(self._data['details'][key])} instead."
```

Comment:
`self._data[...]` or `self._details[...]`? I thought you got rid of `self._data`?